internal/cri/server/stats_collector.go

```diff
@@ -0,0 +1,300 @@
+	if len(sandboxes) == 0 {
+		return
+	}
+
+	timestamp := time.Now()
+	for _, sb := range sandboxes {
+		// Get the parent cgroup path for the pod
+		cgroupPath := sb.Config.GetLinux().GetCgroupParent()
+		if cgroupPath == "" {
+			continue
+		}
+
+		usageCoreNanoSeconds, ok := c.getCgroupCPUUsage(ctx, cgroupPath)
+		if !ok {
+			continue
+		}
+		c.addSample(sb.ID, timestamp, usageCoreNanoSeconds)
+	}
+}
+
+// getCgroupCPUUsage reads CPU usage from a cgroup path.
+// Supports both cgroupv1 and cgroupv2.
+func (c *StatsCollector) getCgroupCPUUsage(ctx context.Context, cgroupPath string) (uint64, bool) {
+	switch cgroups.Mode() {
+	case cgroups.Unified:
+		cg, err := cgroupsv2.Load(cgroupPath)
+		if err != nil {
+			log.G(ctx).WithError(err).Debugf("StatsCollector: failed to load cgroupv2: %s", cgroupPath)
+			return 0, false
+		}
+		stats, err := cg.Stat()
+		if err != nil {
+			log.G(ctx).WithError(err).Debugf("StatsCollector: failed to get cgroupv2 stats: %s", cgroupPath)
+			return 0, false
+		}
+		if stats.CPU != nil {
+			// cgroupv2 reports in microseconds, convert to nanoseconds
+			return stats.CPU.UsageUsec * 1000, true
+		}
+	default:
+		control, err := cgroup1.Load(cgroup1.StaticPath(cgroupPath))
+		if err != nil {
+			log.G(ctx).WithError(err).Debugf("StatsCollector: failed to load cgroupv1: %s", cgroupPath)
+			return 0, false
+		}
+		stats, err := control.Stat(cgroup1.IgnoreNotExist)
+		if err != nil {
+			log.G(ctx).WithError(err).Debugf("StatsCollector: failed to get cgroupv1 stats: %s", cgroupPath)
+			return 0, false
+		}
+		if stats.CPU != nil && stats.CPU.Usage != nil {
+			return stats.CPU.Usage.Total, true
+		}
+	}
+	return 0, false
+}
+
+// addSample adds a CPU sample for the given container/sandbox ID.
+func (c *StatsCollector) addSample(id string, timestamp time.Time, usageCoreNanoSeconds uint64) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	store, ok := c.stores[id]
+	if !ok {
+		store = stats.NewTimedStore(c.statsAge, c.maxSamples)
+		c.stores[id] = store
+	}
+	store.Add(timestamp, usageCoreNanoSeconds)
+}
+
+// GetUsageNanoCores returns the latest calculated UsageNanoCores for the given
+// container/sandbox ID. Returns 0 and false if no data is available or if
+// there aren't enough samples to calculate the rate.
+func (c *StatsCollector) GetUsageNanoCores(id string) (uint64, bool) {
+	c.mu.RLock()
+	store, ok := c.stores[id]
+	c.mu.RUnlock()
+
+	if !ok {
+		return 0, false
+	}
+
+	return store.GetLatestUsageNanoCores()
+}
+
+// GetLatestSample returns the latest CPU sample for the given container/sandbox ID.
+// Returns nil if no data is available.
+func (c *StatsCollector) GetLatestSample(id string) *stats.CPUSample {
+	c.mu.RLock()
+	store, ok := c.stores[id]
+	c.mu.RUnlock()
+
+	if !ok {
+		return nil
+	}
+
+	return store.GetLatest()
+}
+
+// RemoveContainer removes the stats store for a container/sandbox.
+// Should be called when a container is removed.
+func (c *StatsCollector) RemoveContainer(id string) {
```

Comment:
this func is not being called yet when the container is removed from the container store..